Adopt smart pointers in WebFullScreenManagerProxy.cpp, WebInspectorClient.cpp, WebInspectorInternal.cpp

jelee53 opened the pull request:
#### 9774abe04a61e7188ab484149b1f6d23448534a2
<pre>
Adopt smart pointers in WebFullScreenManagerProxy.cpp, WebInspectorClient.cpp, WebInspectorInternal.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=295243">https://bugs.webkit.org/show_bug.cgi?id=295243</a>
<a href="https://rdar.apple.com/154710862">rdar://154710862</a>

Reviewed by Per Arne Vollan, Chris Dumez, and Anne van Kesteren.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::~WebInspectorClient):
(WebKit::WebInspectorClient::openLocalFrontend):
(WebKit::WebInspectorClient::bringFrontendToFront):
(WebKit::WebInspectorClient::didResizeMainFrame):
(WebKit::WebInspectorClient::showPaintRect):
(WebKit::WebInspectorClient::elementSelectionChanged):
(WebKit::WebInspectorClient::timelineRecordingChanged):
(WebKit::WebInspectorClient::setDeveloperPreferenceOverride):
* Source/WebKit/WebProcess/Inspector/WebInspectorInternal.cpp:
(WebKit::WebInspector::~WebInspector):
(WebKit::WebInspector::openLocalInspectorFrontend):
(WebKit::WebInspector::closeFrontendConnection):
(WebKit::WebInspector::bringToFront):
(WebKit::WebInspector::showMainResourceForFrame):
(WebKit::WebInspector::elementSelectionChanged):
(WebKit::WebInspector::timelineRecordingChanged):
(WebKit::WebInspector::setDeveloperPreferenceOverride):
(WebKit::WebInspector::setEmulatedConditions):
(WebKit::WebInspector::canAttachWindow):
(WebKit::WebInspector::updateDockingAvailability):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedInspector):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/296924@main">https://commits.webkit.org/296924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b76e84dc91b15c2dcfea0d5bf93ca783b527446

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20024 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115957 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60182 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30271 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38181 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83573 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24132 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99003 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64015 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23504 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59751 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93503 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17193 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118748 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36974 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92550 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95269 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92374 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23550 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37350 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15095 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32841 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36869 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42340 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36529 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39871 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38238 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -103,22 +103,22 @@ void WebInspectorClient::frontendCountChanged(unsigned count)
-    if (page && page->inspector())
-        page->inspector()->bringToFront();
+    if (page && page->protectedInspector())
+        page->protectedInspector()->bringToFront();
```
```suggestion
    if (!page)
        return;

    if (RefPtr inspector = page->inspector();
        inspector->bringToFront();
```
Is what we did elsewhere for cases like this. You should definitely never need a protected member just for a boolean output.

jelee53 commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -103,22 +103,28 @@ void WebInspectorClient::frontendCountChanged(unsigned count)
 {
     if (RefPtr page = m_page.get())
-        page->inspector()->openLocalInspectorFrontend();
+        page->protectedInspector()->openLocalInspectorFrontend();
```
I'm keeping around the `protectedInspector()` and other protected getters just in case others might need them. I can remove though if we think just one usage in the code doesn't require a helper. 

Thoughts? @annevk @pvollan

annevk replied:
Given it's on `page()` I think that's warranted, might be worth a quick grep to confirm, but seems likely it's needed elsewhere.

annevk commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -189,13 +195,13 @@ void WebInspectorClient::showPaintRect(const FloatRect& rect)
     if (!m_paintRectOverlay) {
         m_paintRectOverlay = PageOverlay::create(*this, PageOverlay::OverlayType::Document);
-        page->corePage()->pageOverlayController().installPageOverlay(*m_paintRectOverlay, PageOverlay::FadeMode::DoNotFade);
+        page->corePage()->pageOverlayController().installPageOverlay(*(m_paintRectOverlay.copyRef()), PageOverlay::FadeMode::DoNotFade);
```
Given that you need `m_paintRectOverlay` again below I'd write this as:

```
RefPtr paintRectOverlay = m_paintRectOverlay;
if (!paintRectOverlay)
    paintRectOverlay = ...
    m_paintRectOverlay = paintRectOverlay.copyRef();

...
... paintRectOverlay->layer();
```
I think that ends up being clearer overall.

jelee53 replied:
I made some stylistic updates to this where paintRectOverlay is not assigned until after m_paintRectOverlay is created. Please let me know what you think!

annevk commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -80,7 +80,7 @@ WebInspectorClient::~WebInspectorClient()
         RefPtr page = m_page.get();
         if (page && page->corePage())
-            page->corePage()->pageOverlayController().uninstallPageOverlay(*m_paintRectOverlay, PageOverlay::FadeMode::Fade);
+            page->corePage()->pageOverlayController().uninstallPageOverlay(*(m_paintRectOverlay.copyRef()), PageOverlay::FadeMode::Fade);
```
You should obtain a `RefPtr` to `m_paintRectOverlay` as part of the conditional above and deref that here.

annevk reviewed (approved):
Modulo `m_paintRectOverlay` nits this looks good to me. (Which I see now are also being flagged by the analyzer as wrong. Oops.)

pvollan commented on Source/WebKit/WebProcess/Inspector/WebInspectorInternal.cpp:
```diff
@@ -101,12 +101,12 @@ void WebInspector::setFrontendConnection(IPC::Connection::Handle&& connectionHan
-        m_frontendConnection = nullptr;
+    if (RefPtr frontendConnection = m_frontendConnection) {
+        frontendConnection->invalidate();
+        frontendConnection = nullptr;
```
Should we set `m_frontendConnection` to nullptr here?

jelee53 replied:
Gosh darn it. I'm so careless. I'll fix that.

pvollan reviewed (approved):
R=me.

annevk commented on Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
```diff
@@ -187,15 +193,17 @@ void WebInspectorClient::showPaintRect(const FloatRect& rect)
         m_paintRectOverlay = PageOverlay::create(*this, PageOverlay::OverlayType::Document);
-        page->corePage()->pageOverlayController().installPageOverlay(*m_paintRectOverlay, PageOverlay::FadeMode::DoNotFade);
+        paintRectOverlay = m_paintRectOverlay;
+        page->corePage()->pageOverlayController().installPageOverlay(*paintRectOverlay, PageOverlay::FadeMode::DoNotFade);
```
I don't think this makes sense as you only assign once. In that case you might as well use a protected method or wrap it inline. But I think the goal should be to make as many of the `m_foo` variables disappear so you only have to track a single variable when reading the code.

annevk replied:
Also, I think because of this you introduced a bug because when `m_paintRectOverlay` is already assigned this code path is not taken and then `paintRectOverlay` is a nullptr causing an issue below.

So please always stick to the rule of replacing all the `m_foo` member variables. That would avoid this kind of issue.

jelee53 replied:
Of course! Yes, `paintRectOverlay` is not initialized in some cases. 

I will fix this. And use `paintRectOverlay` in the places that I can (for example, the if statement).

annevk commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5206,6 +5211,11 @@ WebInspectorUI* WebPage::inspectorUI()
+RefPtr<WebInspectorUI> WebPage::protectedInspectorUI()
+{
+    return inspectorUI();
+}
```
I don't think we should add this if it's not used.

annevk commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5215,6 +5225,11 @@ RemoteWebInspectorUI* WebPage::remoteInspectorUI()
+RefPtr<RemoteWebInspectorUI> WebPage::protectedRemoteInspectorUI()
+{
+    return remoteInspectorUI();
+}
```
Ditto.

annevk reviewed (approved):
Looks great, thanks!

jelee53 committed:
Adopt smart pointers in WebFullScreenManagerProxy.cpp, WebInspectorClient.cpp, WebInspectorInternal.cpp
https://bugs.webkit.org/show_bug.cgi?id=295243
rdar://154710862

Reviewed by Per Arne Vollan, Chris Dumez, and Anne van Kesteren.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/Inspector/WebInspectorClient.cpp:
(WebKit::WebInspectorClient::~WebInspectorClient):
(WebKit::WebInspectorClient::openLocalFrontend):
(WebKit::WebInspectorClient::bringFrontendToFront):
(WebKit::WebInspectorClient::didResizeMainFrame):
(WebKit::WebInspectorClient::showPaintRect):
(WebKit::WebInspectorClient::elementSelectionChanged):
(WebKit::WebInspectorClient::timelineRecordingChanged):
(WebKit::WebInspectorClient::setDeveloperPreferenceOverride):
* Source/WebKit/WebProcess/Inspector/WebInspectorInternal.cpp:
(WebKit::WebInspector::~WebInspector):
(WebKit::WebInspector::openLocalInspectorFrontend):
(WebKit::WebInspector::closeFrontendConnection):
(WebKit::WebInspector::bringToFront):
(WebKit::WebInspector::showMainResourceForFrame):
(WebKit::WebInspector::elementSelectionChanged):
(WebKit::WebInspector::timelineRecordingChanged):
(WebKit::WebInspector::setDeveloperPreferenceOverride):
(WebKit::WebInspector::setEmulatedConditions):
(WebKit::WebInspector::canAttachWindow):
(WebKit::WebInspector::updateDockingAvailability):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::protectedInspector):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/296924@main
